drivers/clock_control/nrf_clock_calibration.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2016-2019 Nordic Semiconductor ASA
+ * Copyright (c) 2016 Vinayak Kariappa Chettimada
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <sensor.h>
+#include <drivers/clock_control/nrf_clock_control.h>
+#include <drivers/clock_control.h>
+#include <nrf_clock.h>
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* For platforms that do not have CTSTOPPED event CT timer can be started
+ * immediately after stop. Redefined events to avoid ifdefs in the code,
+ * CTSTOPPED interrupt handling will be removed during compilation.
+ */
+#ifndef CLOCK_EVENTS_CTSTOPPED_EVENTS_CTSTOPPED_Msk
+#define NRF_CLOCK_EVENT_CTSTOPPED 0
+#endif
+
+#ifndef CLOCK_INTENSET_CTSTOPPED_Msk
+#define NRF_CLOCK_INT_CTSTOPPED_MASK 0
+#endif
+
+#define TEMP_SENSOR_NAME \
+	COND_CODE_1(CONFIG_TEMP_NRF5, (CONFIG_TEMP_NRF5_NAME), (NULL))
+
+/* Calibration state enum */
+enum nrf_cal_state {
+	CAL_OFF,
+	CAL_IDLE,	/* Calibration timer active, waiting for expiration. */
+	CAL_HFCLK_REQ,	/* HFCLK XTAL requested. */
+	CAL_TEMP_REQ,	/* Temperature measurement requested. */
+	CAL_ACTIVE,	/* Ongoing calibration. */
+	CAL_ACTIVE_OFF /* Ongoing calibration, off requested. */
+};
+
+static enum nrf_cal_state cal_state; /* Calibration state. */
+static s16_t prev_temperature; /* Previous temperature measurement. */
+static u8_t calib_skip_cnt; /* Counting down skipped calibrations. */
+
+/* Callback called on hfclk started. */
+static void cal_hf_on_callback(struct device *dev, void *user_data);
+static struct clock_control_async_data cal_hf_on_data = {
+	.cb = cal_hf_on_callback
+};
+
+static struct device *hfclk_dev; /* Handler to hfclk device. */
+static struct device *temp_sensor; /* Handler to temperature sensor device. */
+
+static void measure_temperature(struct k_work *work);
+static K_WORK_DEFINE(temp_measure_work, measure_temperature);
+
+static bool clock_event_check_and_clean(u32_t evt, u32_t intmask)
+{
+	bool ret = nrf_clock_event_check(evt) &&
+			nrf_clock_int_enable_check(intmask);
```

Comment:
Why we care clearing event only if IRQ is enabled?

Reply:
why should we clean it always?

Reply:
Why we are cleaning event only when IRQ is enabled for it?